coder/coder
# feat(scripts): add script to check schema between migrations

- migrations: allow passing in a custom migrate.FS when calling `migrate.Up`
- gen/dump: extract pg_dump used by gen/dump to dbtestutil, replace sed calls
- scripts: write script to test migrations between schema versions

When run on the problematic commit with the gappy migrations:
```
go run ./scripts/migrate-test/main.go --from=bf296b2f1931aa259c573f43d58712f837139c1e --to=main
Read schema at version "main"
Read migrations for "bf296b2f1931aa259c573f43d58712f837139c1e"
Read migrations for "main"
Connect to postgres
No previous migration detected.
Init database at version "bf296b2f1931aa259c573f43d58712f837139c1e"
Migrate to version "main"
Dump schema at version "main"
Schema differs from expected after migration:   (
        """
        ... // 741 identical lines
        );
  
-       CREATE TABLE template_version_parameters (
-           template_version_id uuid NOT NULL,
-           name text NOT NULL,
-           description text NOT NULL,
-           type text NOT NULL,
-           mutable boolean NOT NULL,
-           default_value text NOT NULL,
-           icon text NOT NULL,
-           options jsonb DEFAULT '[]'::jsonb NOT NULL,
-           validation_regex text NOT NULL,
-           validation_min integer,
-           validation_max integer,
-           validation_error text DEFAULT ''::text NOT NULL,
-           validation_monotonic text DEFAULT ''::text NOT NULL,
-           required boolean DEFAULT true NOT NULL,
-           display_name text DEFAULT ''::text NOT NULL,
-           display_order integer DEFAULT 0 NOT NULL,
-           ephemeral boolean DEFAULT false NOT NULL,
-           CONSTRAINT validation_monotonic_order CHECK ((validation_monotonic = ANY (ARRAY['increasing'::text, 'decreasing'::text, ''::text])))
+       CREATE TABLE template_usage_stats (
+           start_time timestamp with time zone NOT NULL,
+           end_time timestamp with time zone NOT NULL,
+           template_id uuid NOT NULL,
+           user_id uuid NOT NULL,
+           median_latency_ms real,
+           usage_mins smallint NOT NULL,
+           ssh_mins smallint NOT NULL,
+           sftp_mins smallint NOT NULL,
+           reconnecting_pty_mins smallint NOT NULL,
+           vscode_mins smallint NOT NULL,
+           jetbrains_mins smallint NOT NULL,
+           app_usage_mins jsonb
        ... // 697 identical, 348 removed, and 408 inserted lines
        """
  )

exit status 1
```

When run on the subsequent commit that fixed it:
```
go run ./scripts/migrate-test/main.go --from=cfa5be52f7c57042690aa60fdbea8098ada93acc --to=main
Read schema at version "main"
Read migrations for "cfa5be52f7c57042690aa60fdbea8098ada93acc"
Read migrations for "main"
Connect to postgres
No previous migration detected.
Init database at version "cfa5be52f7c57042690aa60fdbea8098ada93acc"
Migrate to version "main"
Dump schema at version "main"
OK
```
